src/metaxy/models/feature_spec.py

```diff
@@ -1,16 +1,60 @@
-from collections.abc import Mapping
+from collections.abc import Mapping, Sequence
 from functools import cached_property
+from typing import Any
 
 import pydantic
+from pydantic import field_validator, model_validator
 
 from metaxy.models.field import FieldSpec, SpecialFieldDep
 from metaxy.models.types import FeatureDepMetadata, FeatureKey, FieldKey
 
 
+def _coerce_feature_key(value: Any) -> FeatureKey:
+    """Normalize incoming values to FeatureKey."""
+    if isinstance(value, FeatureKey):
+        return value
+    if isinstance(value, str):
+        return FeatureKey([value])
+    if isinstance(value, Sequence) and not isinstance(value, (str, bytes)):
+        parts = list(value)
+        if not all(isinstance(part, str) for part in parts):
+            raise TypeError("Feature key parts must be strings.")
+        return FeatureKey(parts)
+    raise TypeError("Feature key must be a FeatureKey, string, or sequence of strings.")
+
+
+def _coerce_feature_dep_key(value: Any) -> FeatureKey:
```

Comment:
`hasattr` is a code smell, there should be separate code paths for different types instead. I also think the type annotation here should not be `Any`. There should be a clear `Union` of acceptable types.